[MM-27142] In-Web-App purchase of Mattermost Cloud

nickmisasi opened the pull request:
#### Summary
We want Mattermost Cloud users to be able to purchase cloud licenses without leaving the webapp. This PR adds a modal that can take a credit card, and send payments to a Customer Web Server instance.

In order to see the purchase modal, you can access it through the sidebar main menu, via the button "Purchase Cloud"
When at or over the user limit for the spinwick, as an administrator you will see a banner (either blue or red) with a button to Upgrade Mattermost Cloud. Clicking this will take you to the modal. 
When at or over the user limit, as an administrator, clicking "Invite People" from the main menu will surface a modal, with an "Upgrade Mattermost Cloud" button. Clicking this will take you to the purchase modal.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-27142

#### Related Pull Requests
- Has redux changes https://github.com/mattermost/mattermost-redux/pull/1251

#### Screenshots
![image](https://user-images.githubusercontent.com/12176405/94602693-38a09d00-0263-11eb-82a8-7c944679640f.png)
![image](https://user-images.githubusercontent.com/12176405/94602723-42c29b80-0263-11eb-97b8-1dacc0171f92.png)
![image](https://user-images.githubusercontent.com/12176405/94602744-47874f80-0263-11eb-9628-c4d6504d0b97.png)

marianunez commented on actions/cloud.tsx:
```diff
@@ -0,0 +1,98 @@
+        );
+
+        if (!result) {
+            console.error('Stripe confirm card setup failed.'); //eslint-disable-line no-console
+            return 'Stripe confirm card failed';//'intl.formatMessage({id: 'errors.generic_payment_failure'})'};
```
We still need to add localization for these error messages in this function

nickmisasi replied:
I thought we were saving these for the console log, and showing a generic error to the user? I don't think it's super helpful to tell the end user that we're using stripe. Should I make them more generic? Or just show the same generic error?

marianunez replied:
The console error logs is fine but if we are returning an error here and displaying it somewhere then we need to localize that. If not, then does it make sense to return an error at all? We can return a generic error like the one commented next to it.

nickmisasi replied:
I've changed this function up. It logs the error if it runs into one, but displays generic error to the user. This now returns boolean true/false for success/failure respectively. Let me know what you think now!

marianunez commented on components/cloud/types/customer.ts:
```diff
@@ -0,0 +1,39 @@
+
+//import {LogoutAction} from './user';
+
+export type Customer = {
+    id: string;
+    name: string;
+    email: string;
+    billing_address: Address;
+    company_address: Address;
+    payment_method: PaymentMethod;
+}
+
+export type Address = {
+    city: string;
+    state: string;
+    country: string;
+    postal_code: string;
+    line1: string;
+    line2: string;
+}
+
+export type PaymentMethod = {
+    type: string;
+    last_four: number;
+    exp_month: number;
+    exp_year: number;
+    card_brand: string;
+}
+
+export const SET_CUSTOMER = 'SET_CUSTOMER';
+
+export type SetCustomerAction = {
+    type: typeof SET_CUSTOMER;
+    customer: Customer;
+};
+
+//export type CustomerActionType = SetCustomerAction | LogoutAction;
```
I believe this overlaps with @devinbinnie [`Customer` type definition](https://github.com/devinbinnie/mattermost-redux/blob/727f469505d95b70060c93374adbfe878ab8e951/src/types/cloud.ts#L24) on this Redux PR https://github.com/mattermost/mattermost-redux/pull/1249 . We should use that instead and avoid duplicates in the webapp

devinbinnie replied:
For future reference @nickmisasi, we should try to keep all our types in `mattermost-redux`, especially if they represent data coming from the server.

marianunez replied:
FYI, that PR is merged so we can use that customer type here instead

nickmisasi replied:
Should be good here. I moved the sku's into `types/cloud/sku.ts` but removed this file in favour of Devin's mattermost-redux changes

marianunez commented on components/main_menu/main_menu.jsx:
```diff
@@ -341,6 +343,13 @@ class MainMenu extends React.PureComponent {
                             dialogType={MarketplaceModal}
                             text={formatMessage({id: 'navbar_dropdown.marketplace', defaultMessage: 'Plugin Marketplace'})}
                         />
+                        <Menu.ItemToggleModalRedux
+                            id='cloudPurchaseModal'
+                            modalId={ModalIdentifiers.CLOUD_PURCHASE}
+                            show={!this.props.mobile}
+                            dialogType={PurchaseModal}
+                            text={'Purchase Cloud'}
+                        />
```
This should be removed before merge. This was just a easy access for dev/testing

marianunez commented on components/payment_form/card_input.css:
```diff
@@ -0,0 +1,37 @@
+.StripeElement:focus {
+    border-color: transparent;
+    box-shadow: 0 0 0 2px var(--button-bg);
+}
```
Nit: missing new line

marianunez commented on components/payment_form/payment_form.scss:
```diff
@@ -0,0 +1,93 @@
+        line-height: 20px;
+        color: #0058CC;
+    }
+}
```
Nit: missing new line

marianunez commented on components/payment_form/payment_form.scss:
```diff
@@ -0,0 +1,93 @@
+        font-size: 16px;
+        line-height: 24px;
+        color: #22406D;
+        border-bottom: 1px solid rgba(0, 0, 0, 0.12);
```
We need to change have this use the theme colors but can make it part of the saved card ticket: https://mattermost.atlassian.net/browse/MM-29096

marianunez commented on components/purchase_modal/icon_message.tsx:
```diff
@@ -0,0 +1,96 @@
+            </h3>
+            <div className={classNames('IconMessage-sub', error || '')}>
+                {subtitle ? <FormattedMessage id={subtitle}/> : null}
+            </div>
+            {button}
+            {link}
+            {footer}
```
Do we need to add the error message here? I think it originally had it and I removed it at some point

nickmisasi replied:
Instead of having discrete subtitle and errors in the DOM I just figured an error is a subtitle, but colour it red.

marianunez replied:
Can you post a screenshot of a failure so we can have a UX review on that as well?

marianunez commented on components/purchase_modal/process_payment_setup.tsx:
```diff
@@ -0,0 +1,169 @@
+                <IconMessage
+                    title={t('admin.billing.subscription.upgradedSuccess')}
+                    subtitle={
+                        'Starting August 8, 2020 you will be charged based on the number of enabled users'
```
We need to change this date to use the first date of the next month as we are doing in the modal here https://github.com/mattermost/mattermost-webapp/blob/46247a6f71a288694af34dfa9348d7c2f506536f/components/purchase_modal/purchase_modal.tsx#L83

marianunez commented on components/purchase_modal/purchase.scss:
```diff
@@ -0,0 +1,165 @@
+            top: 0;
+            z-index: -1;
+
+            // this opacity makes it invisible
+            // opacity: 0.16;
```
Should we remove this?

marianunez commented on components/purchase_modal/purchase.scss:
```diff
@@ -0,0 +1,165 @@
+        }
+
+
+        /// ?
+        .SignUpRHS {
+            position: relative;
+            margin-top: 29px;
+            width: 25%;
+            z-index: 2;
+        }
+
+        .SignUpRHS-terms {
+            position: absolute;
+            bottom: 0;
+            opacity: 0.6;
+            font-size: 16px;
+            line-height: 24px;
+            margin-bottom: 42px;
+        }
+
+        .SignUpRHS-link {
+            color: #22406D;
+            margin-left: 32px;
+        }
```
I believe we can remove this, this was part of the migration from CWS but not used here

marianunez commented on images/cloud/upgrade.svg:
```diff
@@ -0,0 +1,30 @@
+<svg width="267" height="163" viewBox="0 0 267 163" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Is this the same one currently used in the user limit modal? It seems we are adding it again

marianunez commented on images/onboarding-bg.svg:
```diff
@@ -0,0 +1,143 @@
+<svg width="1366" height="767" viewBox="0 0 1366 767" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Are these onboarding images intended to be part of this PR? Maybe a bad merge?

devinbinnie replied:
Yeah looks like this overwrites the work we did to theme this.

marianunez reviewed (changes requested):
@nickmisasi awesome work getting this PR complete! Overall looks great! Some comments below

devinbinnie commented:
Looks like we have some duplicate work done with this PR here: https://github.com/mattermost/mattermost-webapp/pull/6543

Not sure which we should keep, but I was hoping to introduce these dropdowns as common components.

marianunez commented:
> Mattermost test server created! 🎉
> 
> Access here: [mattermost-webapp-pr-6583.test.mattermost.cloud](https://mattermost-webapp-pr-6583.test.mattermost.cloud)
> 
> Account Type	Username	Password
> Admin	sysadmin	Sys@dmin123
> User	user-1	User-1@123

@matthewbirtch gentle ping, this is ready for review ASAP given that some items in this PR is blocking some other work on the System Console.

matthewbirtch commented:
Apologies for the dealy @marianunez @nickmisasi. I'll be reviewing tonight

matthewbirtch reviewed (changes requested):
Hey @nickmisasi, first pass review here.

1. Looks like the background image isn't spanning the width of larger screens and is getting cut off on the right
<img width="1151" alt="Screen Shot 2020-09-30 at 8 25 14 PM" src="https://user-images.githubusercontent.com/2040554/94753123-433e5d80-035b-11eb-96fd-6386327ccba2.png">

2. The x icon appears to be slightly off position - it should be equally positioned 12px from the top and right. Right now it is positioned 10px from the top and 20px from the right

3. A Scrollbar seems to be appearing even if it's not needed
![image](https://user-images.githubusercontent.com/2040554/94753615-bd231680-035c-11eb-8c16-5f0af7766510.png)

4. For the illustration element in the left column, I think we need to work on getting the theme colors for the svgs. I believe Asaad had a solution for this and I think we did this for the onboarding stuff too? Right now there are definitely contrast issues with these illustrations
![image](https://user-images.githubusercontent.com/2040554/94755581-fe1e2980-0362-11eb-9583-6140c42e4476.png)

4. For the middle column (form content), rather than setting the width on the form to 75% and max-width of 448px in this case, let's remove those and just add left and right padding of 96px

5. A number of issues with the input fields UI. I just want to make sure you're using the work Devin had done?
    - The credit card input doesn't seem to be using proper theme colors
![image](https://user-images.githubusercontent.com/2040554/94757098-5820ee00-0367-11eb-911d-e2473d5c9457.png)
    - Inputs are supposed to be 40px in height, but the `Name on card`, `Address`, `Address 2`, `City`, `Zip Code` are 36px (or less in the case of the Zip code input)
    - the border color on the select input isn't the same as the text inputs
    - the down arrow icon in the select input doesn't match what Devin did and what's in the designs
    - You'll notice in the dark theme screenshot above, select inputs don't appear to be following theme colors.
![image](https://user-images.githubusercontent.com/2040554/94757146-71c23580-0367-11eb-9ca4-9e4dca3c29b6.png)
    - the labels are not positioned correctly on all inputs
    - the placeholder text color on the inputes seems too light compared to the designs and makes it seem disabled. Although the credit card field placeholder text color seems more right
    - the corner radius doesn't seem quite right on the focus state because the thicker border appears on the outside rather than the inside - can you check with Devin on this? He recently fixed this problem
**Implemented**
![image](https://user-images.githubusercontent.com/2040554/94754540-7edb2680-035f-11eb-99f2-1cc36093b84f.png)
**Designed**
![image](https://user-images.githubusercontent.com/2040554/94754632-c2359500-035f-11eb-9861-7a86b3d0c645.png)

6. The country input don't have the label
![image](https://user-images.githubusercontent.com/2040554/94754415-30c62300-035f-11eb-8972-b991b0a0c9db.png)

7. For select inputs, when the dropdown menu is open, I'm used to being able to type the first few characters of my provice or country and the menu jumps down to it. These select inputs don't do this standard behaviour

8. Need a space in front of 'See how billing works'
![image](https://user-images.githubusercontent.com/2040554/94755797-b21fb480-0363-11eb-8250-bd434b06ada7.png)

9. In the right column summary, can we add `line-height:1` to the `.price-text` line? This should reduce the spacing here to match closer with design

10. Is it possible to anchor the Mattermost Cloud logo to the bottom?

11. Can we use a proper theme colors for the logo as well? We should use the Button BG Color for Mattermost and 48% Center Channel color for the 'Cloud' portion of the logo

12. For the processing screen
    - we need to vertically center the content within the modal
    - we will need to look a theme colors for illustrations and the progress bar
    - it should say "**Verifying** your payment information". Right now it says "Verify..."
![image](https://user-images.githubusercontent.com/2040554/94756361-56eec180-0365-11eb-97c1-36137c565304.png)

13. For the success screen
    - the computer screen illustration should be horizontally centered. It looks like the image is causing a visual off-centered imbalance because of the leaves as part of the image.    
    - it should also be consistently positioned with the previous processing state. It looks like it shifts
    - we need to vertically center the content within the modal
    - we will need to look a theme colors for illustrations
![image](https://user-images.githubusercontent.com/2040554/94756066-79340f80-0364-11eb-8388-fd308fe565ba.png)

7. We'll need to work on the views for smaller screens. A mobile view was designed initially. You can see it here, but it might be beneficial to work together on this. Here's the link to the mobile design for the payment screen: https://www.figma.com/file/uv9A6dZsekQHHhOhYC7yc2/Cloud-UX?node-id=3273%3A106295

This is probably enough for now. Let me know if you would like to hop on a zoom call to go through any of it.

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -0,0 +1,233 @@
+import ProcessPaymentSetup from './process_payment_setup';
+
+const STRIPE_CSS_SRC = 'https://fonts.googleapis.com/css?family=Open+Sans:400,400i,600,600i&display=swap';
+const STRIPE_PUBLIC_KEY = 'pk_test_ttEpW6dCHksKyfAFzh6MvgBj';
```
@joram this is the same as in CWS. Do we need to change this in the build process for the webapp?

marianunez replied:
(Once again I mentioned wrong, apologies `@joram`!)

@jwilander see above

jwilander replied:
Yes, we'll need to replace this when building releases

ali-farooq0 replied:
is this a real key? ahh should this even be here? I propose removing it from the webapp @marianunez

nickmisasi commented:
Thanks again for the detailed evaluation @matthewbirtch 

I believe I've addressed everything in your comment, with the exception of things to do with themeing colours and responsive design. We do plan to do this before GA, through [this ticket](https://mattermost.atlassian.net/browse/MM-29214), but there are a few other core mechanics that are going to take priority. I'll address each of your comments explicitly, to help make sure I've gotten to them all:


> 1. Looks like the background image isn't spanning the width of larger screens and is getting cut off on the right

Should be spanning the entire width of the display now. Just noticed as I posted this screenshot that it looks like since the LHS is 25% of the screen, the middle column is farther away from it. Let me know what you'd like me to do here.
![image](https://user-images.githubusercontent.com/12176405/94929631-594f3980-0493-11eb-9da5-4e4e277c99d0.png)


> 2. The x icon appears to be slightly off position - it should be equally positioned 12px from the top and right. Right now it is positioned 10px from the top and 20px from the right

Should be good here (see image above)

> 3. A Scrollbar seems to be appearing even if it's not needed

Wasn't really able to reproduce this, but I added a CSS attribute to make sure it won't happen anyways



> 5. For the middle column (form content), rather than setting the width on the form to 75% and max-width of 448px in this case, let's remove those and just add left and right padding of 96px

Done




> A number of issues with the input fields UI. I just want to make sure you're using the work Devin had done?

Didn't actually have Devin's dropdowns, since they weren't yet merged. He has merged them now, and I've integrated them for the country and state selector.



> * the labels are not positioned correctly on all inputs
> * the placeholder text color on the inputes seems too light compared to the designs and makes it seem disabled. Although the credit card field placeholder text color seems more right
> * the corner radius doesn't seem quite right on the focus state because the thicker border appears on the outside rather than the inside - can you check with Devin on this? He recently fixed this problem

Think I've got the labels in the correct position now. A little hard to decipher exactly where to place from Figma. Let me know.
I believe the placeholder text is good to go
Corner radius should be fixed by the inset borders



> 7. The country input don't have the label
> 8. For select inputs, when the dropdown menu is open, I'm used to being able to type the first few characters of my provice or country and the menu jumps down to it. These select inputs don't do this standard behaviour

Fixed by integrating Devin's dropdown

> 9. Need a space in front of 'See how billing works'

Pesky HTML. Fixed.

> 10. In the right column summary, can we add line-height:1 to the .price-text line? This should reduce the spacing here to match closer with design

Added. I didn't see any visual change though.

> 11. Is it possible to anchor the Mattermost Cloud logo to the bottom?

Anchored.







> For the processing screen

> * we need to vertically center the content within the modal

Centered

> * we will need to look a theme colors for illustrations and the progress bar

(will be done later)

> * it should say "**Verifying** your payment information". Right now it says "Verify..."

Fixed



> For the success screen

> * the computer screen illustration should be horizontally centered. It looks like the image is causing a visual off-centered imbalance because of the leaves as part of the image.

This one was a little challenging because of the leaves. Let me know how it looks, I can tweak it slightly if you think it needs it.

> * it should also be consistently positioned with the previous processing state. It looks like it shifts

I think I managed this as part of the above.

> * we need to vertically center the content within the modal

Should be good here

matthewbirtch reviewed (approved):
@nickmisasi there are a few tweaks to make follow-up tickets for (see below). However, I'm going to approve as they are not blockers.

1. Credit card input doesn't seem to have a focus state
2. Credit card input label animates off strangely when I leave focus on the field
![2020-10-02 16 39 21](https://user-images.githubusercontent.com/2040554/94968072-40b04500-04ce-11eb-8218-ab741b969f23.gif)
3. The border seems to shift in the focus state (I had this issue in one of Devin's PRs too) and it seems to be adding a white border around the outside (you can see it more against the wave background image. 
![2020-10-02 16 40 06](https://user-images.githubusercontent.com/2040554/94968198-7f45ff80-04ce-11eb-8389-ce48fa23200f.gif)
![image](https://user-images.githubusercontent.com/2040554/94968324-b1576180-04ce-11eb-858a-ab005b1b3380.png)
4. The underline includes the space on the 'See how billing works' hover state
![image](https://user-images.githubusercontent.com/2040554/94968449-ecf22b80-04ce-11eb-893f-95cd80b8bc20.png)
5. Not sure there will be anything we can do about this, but when I use Chrome's autofill on the form, it highlights the fields that it fills, but the highlight doesn't seem to span the full input box oddly enough
![image](https://user-images.githubusercontent.com/2040554/94968580-30e53080-04cf-11eb-9d7b-18c697005f51.png)

nickmisasi commented:
@matthewbirtch I've fixed 4, and I think I fixed 3 and 5 also. 1, and 2, I can note down and address as part of the theming, the Stripe card input is very hard to get styles to apply to.

nickmisasi commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -0,0 +1,233 @@
+                                    className='waves'
+                                    src={wavesBackground}
+                                />
+                                <img
+                                    className='blue-dots'
+                                    src={blueDotes}
+                                />
```
These are the blue dots in the top left of the modal: 
![image](https://user-images.githubusercontent.com/12176405/95116798-94609f80-0715-11eb-919d-d8c37b41422d.png)

I don't see dots in the bottom right on the designs - could you link?

marianunez replied:
<img width="411" alt="Screen Shot 2020-10-05 at 2 20 02 PM" src="https://user-images.githubusercontent.com/936315/95117077-f7eacd00-0715-11eb-8381-d7029b4e77bf.png">

nickmisasi replied:
Totally couldn't see that in the designs until I zoomed in

I've added them - the positioning figma spit out wasn't very good, and the opacity there of 0.08 made it invisible on my screen - so I tweaked it slightly. Hope that's ok!

devinbinnie commented on components/payment_form/state_selector.tsx:
```diff
@@ -0,0 +1,65 @@
+                    value: stateCode.code,
+                    label: stateCode.name,
+                }))}
+                legend={'State/Province'}
```
These should be translated.

devinbinnie commented on components/purchase_modal/icon_message.scss:
```diff
@@ -0,0 +1,94 @@
+            margin-top: 90px;
+            height: 4px;
+            width: 387px;
+            background: rgba(34, 64, 109, 0.12);
```
Themeing?

nickmisasi replied:
Theming will be fixed as part of: https://mattermost.atlassian.net/browse/MM-29214

devinbinnie commented on components/purchase_modal/index.js:
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
This should be a `.ts` file.

devinbinnie commented on components/purchase_modal/process_payment.css:
```diff
@@ -0,0 +1,19 @@
+    margin-top: 60px;
+    height: 4px;
+    width: 387px;
+    background: rgb(34,64,109, 0.12);
```
Does this need themeing?

nickmisasi replied:
Same here

devinbinnie commented on components/purchase_modal/purchase.scss:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        .full-width {
+            border-color: var(--center-channel-color-16)
```
Same here with the use of `--center-channel-color-rgb`

nickmisasi replied:
fixed

devinbinnie commented on images/onboarding-success.svg:
```diff
@@ -0,0 +1,40 @@
+<svg width="444" height="313" viewBox="0 0 444 313" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Same with this file

devinbinnie commented on utils/utils.jsx:
```diff
@@ -1895,3 +1897,8 @@ export function adjustSelection(inputBox, e) {
 }
+
+export function getNextBillingDate() {
+    const nextBillingDate = moment().add(1, 'months').startOf('month');
```
If this is always the start of the month, wouldn't that make the date always the 1st?
If so, couldn't we just use the normal `Date` object?
```suggestion
    const nextBillingDate = new Date();
    nextBillingDate.setMonth(nextBillingDate.getMonth() + 1);
    nextBillingDate.setDay(1);
```

I only say that because you generally don't need `moment.js` for most things. They even say that themselves:
https://momentjs.com/docs/#/-project-status/

nickmisasi replied:
Couple reasons I'd prefer to keep moment as is:
1. Formatting the date is cleaner
2. Moment usually does some magic in the background around localization with its formatting. Things beyond just "translate December to french", but also how the month/day/year is ordered

devinbinnie replied:
Okay, we can leave as-is for now. Might work on removing `moment.js` throughout the app at some point. But i'm only 1/5 on that.

nickmisasi commented on components/payment_form/card_image.tsx:
```diff
@@ -0,0 +1,54 @@
+
+function getCardImage(brand: string): string {
+    switch (brand) {
+    case 'amex':
```
It's not returning a string, but a component (which is the cards image). '' means you can check for if it exists or not, and conditionally render the card (or not)

If you returned brand, this check would break. It could be anything falsey, but can't be truthy (ie, brand)

nickmisasi commented on components/payment_form/card_input.tsx:
```diff
@@ -0,0 +1,185 @@
+    complete: boolean;
+}
+
+const REQUIRED_FIELD_TEXT = 'This field is required';
```
These actually are localized before being displayed

nickmisasi commented on components/payment_form/card_input.tsx:
```diff
@@ -0,0 +1,185 @@
+        if (required && empty) {
+            error = REQUIRED_FIELD_TEXT;
+
+            //t('shortcuts.header'),
```
Nope

nickmisasi commented on components/cloud/utils/states.ts:
```diff
@@ -0,0 +1,79 @@
+    name: string;
+}
+
+export const US_STATES = [
```
There was no folder for `constants` in utils, so I moved the `states.ts` file into `utils` folder. Couldn't put in utils/constants.jsx because typescript and

nickmisasi commented on components/payment_form/payment_form.tsx:
```diff
@@ -0,0 +1,334 @@
+
+    public componentDidUpdate(prevProps: Props) {
+        if (prevProps.paymentMethod == null && this.props.paymentMethod != null) {
+            this.setState(this.getResetState()); //eslint-disable-line react/no-did-update-set-state
```
getDerivedStateFromProps doesn't have access to the previous props, which is needed in these cases. React docs for getDerivedStateFromProps say:
```
If you need to perform a side effect in response to a change in props, use componentDidUpdate lifecycle instead.
```
I think the side-effect here is clearing the form. 

I've tweaked how this is so we aren't eslint ignoring things anymore - let me know what you think!

nickmisasi commented on actions/cloud.tsx:
```diff
@@ -0,0 +1,99 @@
+            cloudProducts = await Client4.getCloudProducts();
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(`Error fetching cloud products: ${error}`);
```
These were actually debugging ones - I don't think these should be shared to the console/user since its around payments. You're right, they'll be put on the redux error stack.

nickmisasi commented on actions/cloud.tsx:
```diff
@@ -0,0 +1,99 @@
+
+import {StripeSetupIntent, BillingDetails} from 'types/cloud/sku';
+
+export function getProductPrice() {
```
Good callout. Should be fixed in newest diff

devinbinnie commented on components/payment_form/payment_form.tsx:
```diff
@@ -0,0 +1,331 @@
+
+type Props = {
+    className: string;
+    initialBillingDetails?: BillingDetails | null;
```
Since, this is an optional parameter, the type evaluates to `BillingDetails | null | undefined`.
For simplicity, could we remove the null and just check for `undefined` or just do `!this.props.initialBillingDetails`

devinbinnie commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -0,0 +1,252 @@
+    actions: {
+        closeModal: () => void;
+        getCloudProducts: () => void;
+        completeStripeAddPaymentMethod: (stripe: Stripe, billingDetails: BillingDetails, isDevMode: boolean) => any;
```
The return type should be something like `boolean | ServerError`

nickmisasi replied:
Think we're good here now!

devinbinnie reviewed (approved):
Just a couple TS nit-picks but otherwise this is looking good!
Nice work @nickmisasi!

marianunez commented:
@stevemudie this is ready for `QA Review`, I've added some additional QA steps in the epic: https://mattermost.atlassian.net/browse/MM-27142

stevemudie reviewed (approved):
Worked with Nick and muscled through the tests we could do on it.
Looks good and will of course test again once it's fully in place.

nickmisasi commented:
/cherry-pick cloud-ga

mattermod commented:
Cherry pick is scheduled.
